File: services/api/src/main/java/net/blueshell/api/service/brevo/ContactService.java
package net.blueshell.api.service.brevo;

import net.blueshell.api.mapping.BrevoContactMapper;
import net.blueshell.api.model.ContributionPeriod;
import net.blueshell.api.model.User;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import sendinblue.ApiClient;
import sendinblue.ApiException;
import sendinblue.Configuration;
import sendinblue.auth.ApiKeyAuth;
import sibApi.ContactsApi;
import sibModel.*;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Service
public class ContactService {

    @Autowired
    private BrevoContactMapper brevoContactMapper;

    @Value("${brevo.apiKey}")
    private String apiKey;
    @Value("${brevo.folders.contributionPeriodsId}")
    private Long contributionPeriodsFolder;

    private ContactsApi getContactsApi() {
        ApiClient defaultClient = Configuration.getDefaultApiClient();
        ApiKeyAuth apiKeyAuth = (ApiKeyAuth) defaultClient.getAuthentication("api-key");
        apiKeyAuth.setApiKey(this.apiKey);
        return new ContactsApi();
    }

    // Retrieves a User from Brevo without overwriting non-null values.
    public User getUserFromBrevo(String email) throws ApiException {
        ContactsApi api = getContactsApi();
        GetExtendedContactDetails details = api.getContactInfo(email, null, null);
        // Create a new User instance (or use an existing one) and map attributes.
        User user = new User();
        Map<String, Object> attributes = (Map<String, Object>) details.getAttributes();
        brevoContactMapper.attributesToUser(attributes, user);
        return user;
    }

    public User getUpdate(User user) throws ApiException {
        if (user.getContactId() != null) {
            return user;
        }
        try {
            ContactsApi api = getContactsApi();
            GetExtendedContactDetails details = api.getContactInfo(user.getEmail(), null, null);
            user.setContactId(details.getId());
        } catch (ApiException e) {
            e.printStackTrace();
        }
        return user;
    }

    public void updateContact(User user) throws ApiException {
        user = getUpdate(user);
        if (user.getContactId() != null) {
            sendUpdate(user);
        } else {
            createContact(user);
        }
    }

    private void createContact(User user) throws ApiException {
        ContactsApi api = getContactsApi();
        CreateContact createContact = new CreateContact();
        createContact.setEmail(user.getEmail());
        // Use the mapper to create the attributes map (null values are filtered out).
        Map<String, Object> attributes = brevoContactMapper.userToAttributes(user);
        createContact.setAttributes(attributes);
        CreateUpdateContactModel response = api.createContact(createContact);
        user.setContactId(response.getId());
    }

    private void sendUpdate(User user) throws ApiException {
        ContactsApi api = getContactsApi();
        UpdateContact updateContact = new UpdateContact();
        // Again, use the mapper so that only non-null attributes are updated.
        updateContact.setAttributes(brevoContactMapper.userToAttributes(user));
        api.updateContact(user.getContactId().toString(), updateContact);
    }

    public void createContributionPeriodList(ContributionPeriod contributionPeriod) throws ApiException {
        ContactsApi api = getContactsApi();
        CreateList createList = new CreateList();
        String periodName = String.format("Contribution Paid %d - %d",
                contributionPeriod.getStartDate().getYear(), contributionPeriod.getEndDate().getYear());
        createList.name(periodName);
        createList.setFolderId(contributionPeriodsFolder);
        CreateModel createModel = api.createList(createList);
        contributionPeriod.setListId(createModel.getId());
    }

    public void addToContributionPeriodList(ContributionPeriod contributionPeriod, User user) throws ApiException {
        ContactsApi api = getContactsApi();
        List<Long> ids = new ArrayList<>();
        ids.add(user.getContactId());
        AddContactToList contactEmails = new AddContactToList();
        contactEmails.setIds(ids);
        api.addContactToList(contributionPeriod.getListId(), contactEmails);
    }

    public void removeFromContributionPeriodList(ContributionPeriod contributionPeriod, User user) throws ApiException {
        ContactsApi api = getContactsApi();
        List<Long> ids = new ArrayList<>();
        ids.add(user.getContactId());
        RemoveContactFromList contactEmails = new RemoveContactFromList();
        contactEmails.setIds(ids);
        api.removeContactFromList(contributionPeriod.getListId(), contactEmails);
    }
}

File: api/src/main/kotlin/net/blueshell/api/platform/integration/contact/ContactService.kt
package net.blueshell.api.platform.integration.contact

import net.blueshell.api.shared.enums.Role
import net.blueshell.api.feature.user.model.User
import net.blueshell.api.feature.contribution.model.ContributionPeriod
import net.blueshell.api.feature.user.service.UserService
import net.blueshell.clients.brevo.api.ContactsApi
import net.blueshell.clients.brevo.model.*
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import org.springframework.web.client.HttpClientErrorException
import org.springframework.web.client.RestClient
import org.springframework.web.client.RestClientResponseException

@Service
class ContactService(
    private val users: UserService,
    private val restClientBuilder: RestClient.Builder,
    @param:Value($$"${brevo.apiKey}") private val apiKey: String,
    @param:Value($$"${brevo.baseUrl:https://api.brevo.com/v3}") private val brevoBaseUrl: String,
    @param:Value($$"${brevo.folders.contributionPeriodsId}") private val contributionPeriodsFolder: Long,
) {
    private val contactsApi: ContactsApi
        get() {
            val client = restClientBuilder
                .baseUrl(brevoBaseUrl)
                .defaultHeader("api-key", apiKey)
                .build()
            return ContactsApi(client)
        }

    fun getUpdate(user: User) {
        if (user.contactId != null) return
        log.info("Getting update for user: {}", user.email)

        try {
            val api = this.contactsApi
            val details =
                api.getContactInfo(user.email, "email_id", null, null)
            user.contactId = details.id
        } catch (_: HttpClientErrorException) {
            log.info("Failed to get contact details for user: {}", user.email)
        }
    }

    fun sync(user: User) {
        getUpdate(user)
        if (user.contactId != null) {
            sendUpdate(user)
        } else {
            createContact(user)
        }
    }

    @Throws(RestClientResponseException::class)
    private fun createContact(user: User) {
        log.info("Creating contact for user: {}", user.email)
        val api = this.contactsApi
        val createContact = toCreateContact(user)
        val response = api.createContact(createContact)
        users.updateContactId(user.id!!, response.id!!)
    }

    @Throws(RestClientResponseException::class)
    private fun sendUpdate(user: User) {
        log.info("Sending update for user: {}", user.email)
        val api = this.contactsApi
        val contact = toUpdateContact(user)
        api.updateContact(
            user.email,
            contact,
            "email_id"
        )
    }

    private fun toCreateContact(user: User): CreateContact {
        val contact = CreateContact()
        contact.email = user.email
        contact.extId = user.id.toString()
        contact.attributes = toAttributes(user)
        return contact
    }

    private fun toUpdateContact(user: User): UpdateContact {
        val contact = UpdateContact()
        contact.extId = user.id.toString()
        contact.attributes = toAttributes(user)
        return contact
    }

    private fun toAttributes(user: User): MutableMap<String, Any> {
        val attrs: MutableMap<String, Any> = HashMap()
        attrs["NEWSLETTER"] = user.newsletter
        attrs["IS_MEMBER"] = user.hasRole(Role.MEMBER)
        attrs["FIRSTNAME"] = user.firstName
        attrs["LASTNAME"] = user.lastName
        attrs["SURNAME"] = user.lastName
        attrs["SMS"] = user.phoneNumber!!
        attrs["WHATSAPP"] = user.phoneNumber!!
        return attrs
    }

    @Throws(RestClientResponseException::class)
    fun createList(contributionPeriod: ContributionPeriod): Long {
        if (contributionPeriod.listId != null) {
            return contributionPeriod.listId!!
        }

        val api = this.contactsApi

        val periodName = String.format(
            "Contribution Paid %d - %d",
            contributionPeriod.startDate.year,
            contributionPeriod.endDate.year
        )
        val createList = CreateList(
            periodName,
            contributionPeriodsFolder
        )
        val createModel = api.createList(createList)
        return createModel.id
    }

    @Throws(RestClientResponseException::class)
    fun addToList(contributionPeriod: ContributionPeriod, user: User) {
        if (user.contactId == null) {
            sync(user)
        }

        val api = this.contactsApi
        val ids: MutableList<Long> = ArrayList()
        ids.add(user.contactId!!)
        val payload = AddContactToListRequest()
        payload.ids = ids
        api.addContactToList(contributionPeriod.listId!!, payload)
    }

    @Throws(RestClientResponseException::class)
    fun removeFromList(contributionPeriod: ContributionPeriod, contactId: Long) {
        val api = this.contactsApi
        val ids: MutableList<Long> = ArrayList()
        ids.add(contactId)
        val payload = RemoveContactFromListRequest()
        payload.ids = ids
        api.removeContactFromList(contributionPeriod.listId!!, payload)
    }

    companion object {
        private val log = LoggerFactory.getLogger(ContactService::class.java)
    }
}
